Source/MQTTnet/Client/MqttClient.cs

```diff
@@ -222,8 +222,11 @@ private void ThrowIfConnected(string message)
 
             try
             {
-                await WaitForTaskAsync(_packetReceiverTask, sender).ConfigureAwait(false);
-                await WaitForTaskAsync(_keepAliveMessageSenderTask, sender).ConfigureAwait(false);
+                if (clientWasConnected)
```

Comment:
this condition also works instead of exception type checking

Reply:
When server is closed,the SendKeepAliveMessagesAsync and ReceivePacketsAsync will all catch exceptions ,then then will all call the ```DisconnectInternalAsync``` method,in ```DisconnectInternalAsync``` method,under code whill deadlock
 await WaitForTaskAsync(_packetReceiverTask, sender).ConfigureAwait(false);
 await WaitForTaskAsync(_keepAliveMessageSenderTask, sender).ConfigureAwait(false);

I add under code,then the Disconnected event whill be call
![image](https://user-images.githubusercontent.com/7851430/42488482-5999ed00-8438-11e8-87b2-18266788cb3d.png)

my English is not good,please forgive me ~

Reply:
But if the disconnect i because of server crash/stop, the Disconnected event is called twice ...

Reply:
I have find the reason, You can see #358